_source/_posts/2019-12-20-five-tools-improve-java.adoc

```diff
@@ -0,0 +1,345 @@
+---
+layout: blog_post
+title: "Five Tools to Improve Your Java Code"
+author: bdemers
+description: "Five great tools to reduce bugs and make your Java code more secure!"
+tags: [java, ci, builds, secure]
+tweets:
+- "These 5 Java tools will help you reduce bugs!"
+- "Test coverage, static analysis and more, 5 tools every Java developer should know!"
+- "Improve your Java code quality the easy way, automate it!"
+image: blog/five-tools-improve-java/stop-bad-code.png
+---
+
+:toc: macro
+:page-liquid:
+:experimental:
+
+Writing quality code takes practice. To write better code, you need to know what should improve. Code quality and what makes code easy to read are very subjective; ask five different developers, you will get six different answers. For this post, I'll avoid most of the subjective and focus on ways to detect real issues and potential bugs.
+
+I wrote some intentionally bad code to demo these tools (which was harder than you might think). If you cannot wait and just want to see the code, you can grab it on https://github.com/oktadeveloper/okta-java-tools-example[GitHub]. Clone the project and run `./mvnw verify` and you will see the issues.
+
+First, it's always a good idea to start with tests!
+
+toc::[]
+
+== 1. Add Test Coverage with JaCoCo
+
+Many projects aim for 100% test coverage. This metric can be deceiving, a low coverage score is generally a bad sign, but a high score doesn't say anything about the quality of tests. There is a point of diminishing returns where you start writing tests just to hit an arbitrary coverage number. You should instead use test coverage to find problems and give you confidence in the quality of your code.
+
+There are a few coverage options in Java-land, but my favorite has been https://www.eclemma.org/jacoco/[JaCoCo] (originally named EclEmma, which is still in the URL).
+
+JaCoCo runs as a Java Agent, which makes it easy to add to any test framework (or any `java` execution). For Maven-based projects you simply add the https://www.eclemma.org/jacoco/trunk/doc/maven.html"[JaCoCo Maven Plugin] to your `poml.xml`:
+
+.pom.xml
```

Comment:
I don't think it's necessary to provide file names when it's obvious. I like to use it when I specify the filename in text, then the full path here.